### DISCUSSION TITLE
What's the best way to handle long running task without blocking the UI?

### DISCUSSION THREAD

**Discussion by gtamas** [Relm4/Relm4#547] [General]
So I have SimpleComponent. In its `update() ` method, I have a message handler which spawns a child process to run some task.
This is kinda a long running task, so I would like to show a spinner while it's running, then when it's done,  hide the spinner and show a bunch of other widgets (form in gtk:Box)

I know there are multiple ways to handle this with Relm4 / Rust, but since I'm new to both I'm not sure what's the best approach. What do I need here? A worker? Async update function? Something else? 

I checked this page: https://aaronerhardt.github.io/relm4-book/book/threads_and_async.html but it's not clear to me how could I implement this using SimpleComponent or something similar.  Any help would be appreciated.

Basically I just wanna run something like this in a non-blocking way:

```
 fn update(&mut self, message: Self::Input, sender: ComponentSender<Self>) {
        match message {
            Something::Execute => {
               self.executing = true
                let mut cmd = Command::new("command goes here..");
                ...
                let out = cmd.output().expect("Command failed to start");
                self.executing = false;
                self.output_buf
                    .set_text(&String::from_utf8_lossy(&out.stdout));
                self.error_buf
                    .set_text(&String::from_utf8_lossy(&out.stderr));
            }
        }
    }
```

**Comment by AaronErhardt** [comment]
> https://aaronerhardt.github.io/relm4-book/book/threads_and_async.html

That's a really old version of the book. I should probably take it down. How did you find this btw?

Please head over to https://relm4.org/book/stable for an up-to-date version of the book. It also has a threads and async chapter with an [overview](https://relm4.org/book/stable/threads_and_async/overview.html) that should point you in the right direction. If you still have questions, feel free to comment in this discussion or ask in our Matrix room :)

**Reply by gtamas** [reply]
Thank you for the answer!

I checked the new book and the documentation was very clear. I implemented this using Commands.
